satpy/readers/agrl_l1.py

```diff
@@ -0,0 +1,178 @@
+
+        data.attrs.update({'platform_name': self['/attr/Satellite Name'],
+                           'sensor': self['/attr/Sensor Identification Code']})
+        data.attrs.update(ds_info)
+
+        return data
+
+    def get_area_def(self, key):
+        # Coordination Group for Meteorological Satellites LRIT/HRIT Global Specification
+        # https://www.cgms-info.org/documents/cgms-lrit-hrit-global-specification-(v2-8-of-30-oct-2013).pdf
+        res = key.resolution
+        coff = _COFF_list[_resolution_list.index(res)]
+        loff = _LOFF_list[_resolution_list.index(res)]
+        cfac = _CFAC_list[_resolution_list.index(res)]
+        lfac = _LFAC_list[_resolution_list.index(res)]
+        a = 6378137.0   # equator radius (m)
+        b = 6356752.3   # polar radius(m)
+        H = 42164000.0  # the distance between spacecraft and centre of earth (m)
+        h = H - a       # the altitude of satellite
+
+        lon_0 = self.file_content['/attr/NOMCenterLon']
+        nlines = self.file_content['/attr/RegLength']
+        ncols = self.file_content['/attr/RegWidth']
+
+        cols = 0
+        left_x = (cols - coff) * (2.**16 / cfac)
+        cols += ncols - 1
+        right_x = (cols - coff) * (2.**16 / cfac)
+
+        lines = self.file_content['/attr/Begin Line Number']
+        upper_y = -(lines - loff) * (2.**16 / lfac)
+        lines = self.file_content['/attr/End Line Number']
+        lower_y = -(lines - loff) * (2.**16 / lfac)
+        area_extent = (np.deg2rad(left_x) * h, np.deg2rad(lower_y) * h,
+                       np.deg2rad(right_x) * h, np.deg2rad(upper_y) * h)
+
+        proj_dict = {'a': float(a),
+                     'b': float(b),
+                     'lon_0': float(lon_0),
+                     'h': float(h),
+                     'proj': 'geos',
+                     'units': 'm'}
+
+        area = geometry.AreaDefinition(
+            self.filename_info['observation_type'],
+            "AGRL {} area".format(self.filename_info['observation_type']),
+            'FY-4A',
+            proj_dict,
+            ncols,
+            nlines,
+            area_extent)
+
+        self.area = area
+
+        return area
+
+    def dn2reflectance(self, dn, slope, offset):
+        """Convert digital number (DN) to reflectance
+        Args:
+            dn: Raw detector digital number
+            slope: Slope
+            offset: Offset
+        Returns:
+            Reflectance [%]
+        """
+        ref = dn * slope + offset
+        ref *= 100  # set unit to %
+
+        return ref.clip(min=0)
+
+    def calibrate(self, data, cal_key):
+        """Calibrate digital number (DN) to brightness_temperature
+        Args:
+            dn: Raw detector digital number
+            cal_key: the name of CAL variable
+        Returns:
+            brightness_temperature [K]
+        """
+        # get indexes of nan values
+        nan_index = np.isnan(data).values
+        # assign 0 to nan values temporarily

```

Comment:
This is converting the dataset to a bumpy array, which we want to avoid. I see that the calibration uses a look up table, check here how to do it without converting: https://github.com/pytroll/satpy/blob/5bfcda85bb8f282c446c5d320f86f1a62afd61e3/satpy/readers/goes_imager_hrit.py#L430